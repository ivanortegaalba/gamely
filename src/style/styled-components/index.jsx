export { theme } from './theme'

// Components
export { Header } from './components/Header'
export { Button } from './components/Button'

// Elements
export {H1, H2, H3, H4, H5, H6} from './elements/Headings'

// Settings
export { colors } from './settings/colors'
export { fonts } from './settings/fonts'

// Tools
export { px2rems } from './tools/functions'
