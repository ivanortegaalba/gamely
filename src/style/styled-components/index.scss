@import '~sanitize.css';

// Settings
@import 'settings/colors';
@import 'settings/fonts';

// Tools
@import 'tools/functions';

// Components
@import 'components/button';
@import 'components/input';
@import 'components/badge';
@import 'components/table-list';
@import 'components/form-control';
@import 'components/header';

// Elements
@import 'elements/headings';

// Objects
@import 'objects/grid';
@import 'objects/spacing';
@import 'objects/flex';

body {
  font-family: $base-font-family;
  -webkit-font-smoothing: antialiased;
}
