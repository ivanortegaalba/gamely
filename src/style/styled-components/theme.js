import { colors } from './settings/colors';
import { fonts } from './settings/fonts';
import { spacing } from './settings/spacing';
import { fontSizes } from './settings/fontSizes';

export const theme = {
  fonts: { ...fonts },
  palette: { ...colors },
  spacing: { ...spacing },
  fontSizes: { ...fontSizes },
};
