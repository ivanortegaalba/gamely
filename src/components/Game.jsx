import React from 'react'
import PropTypes from '../PropTypes'
import { Button } from '../style/styled-components'
import { GAME_LIBRARY_URL } from '../constants'

export function Game ({ game, addToFavourites, removeFromFavourites }) {
  const getLogoUrl = ({ short }) => `${GAME_LIBRARY_URL}/images/games/${short}/${short}_60x60.gif`
  const getGameURL = ({ url }) => `${GAME_LIBRARY_URL}${url}`
  const onClickFavourite = () => {
    if (!game.isFavourite) {
      addToFavourites(game.short)
    }
    else {
      removeFromFavourites(game.short)
    }
  }
  const goToPlay = () => {window.open(getGameURL(game), '_blank')}
  return <article>
    <figure>
      <img src={getLogoUrl(game)} alt='The Pulpit Rock' width='60' height='60' />
      <figcaption>{game.name}</figcaption>
    </figure>
    <footer>
      {game.isFavourite ? <Button
          dataSpec={'remove-button'}
          palette={'danger'}
          size={'s'}
          onClick={onClickFavourite}
          margin={'0 xs 0 0'}
          reverse>
          Remove
        </Button>
        : <Button
          dataSpec={'save-button'}
          palette={'primary'}
          size={'s'}
          onClick={onClickFavourite}
          margin={'0 xs 0 0'}
        >
          Save
        </Button>}
      <Button
        dataSpec={'play-button'}
        size={'s'}
        margin={'0 0 0 xs'}
        onClick={goToPlay}
        reverse={true}>
        Play
      </Button>
    </footer>
  </article>
}

Game.propTypes = {
  game: PropTypes.game.isRequired,
  addToFavourites: PropTypes.func.isRequired,
  removeFromFavourites: PropTypes.func.isRequired
}
Game.defaultProps = {}

export default Game
