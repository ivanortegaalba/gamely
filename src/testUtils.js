import 'jest-styled-components';
import React from 'react';
import { shallow } from 'enzyme';
import { ThemeProvider } from 'styled-components';

export const shallowWithTheme = (component, theme) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return shallow(component, { context });
};

export const mountWithTheme = (component, theme) => {
  const context = mount(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return mount(component, { context });
};

export const shallowWithStore = (component, store) => {
  const context = {
    store,
  };
  return shallow(component, { context });
};
