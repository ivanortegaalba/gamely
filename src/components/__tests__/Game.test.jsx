import React from 'react';
import Game from '../Game';
import mockGame from '../../__mocks__/game';
import { shallowWithTheme } from '../../testUtils';
import { theme } from '../../style/styled-components/theme';

describe('<Game/>', () => {
  describe('represent a no favourite game', () => {
    const removeFromFavourites = jest.fn();
    const addToFavourites = jest.fn();
    const component = (
      <Game
        game={mockGame}
        removeFromFavourites={removeFromFavourites}
        addToFavourites={addToFavourites}
      />
    );
    const wrapper = shallowWithTheme(component, theme);
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should hide the remove button', () => {
      expect(wrapper.find({ dataSpec: 'remove-button' })).toHaveLength(0);
    });
    it('should call addToFavourites when the save button is clicked', () => {
      wrapper.find({ dataSpec: 'save-button' }).simulate('click');
      expect(addToFavourites).toHaveBeenCalledTimes(1);
      expect(removeFromFavourites).toHaveBeenCalledTimes(0);
      expect(addToFavourites).toHaveBeenCalledWith(mockGame.short);
    });
  });
  describe('represent a favourite game', () => {
    const removeFromFavourites = jest.fn();
    const addToFavourites = jest.fn();
    const component = (
      <Game
        game={{ ...mockGame, isFavourite: true }}
        removeFromFavourites={removeFromFavourites}
        addToFavourites={addToFavourites}
      />
    );
    const wrapper = shallowWithTheme(component, theme);
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should hide the remove button', () => {
      expect(wrapper.find({ dataSpec: 'save-button' })).toHaveLength(0);
    });
    it('should call removeFromFavourites when the remove button is clicked', () => {
      wrapper.find({ dataSpec: 'remove-button' }).simulate('click');
      expect(removeFromFavourites).toHaveBeenCalledTimes(1);
      expect(addToFavourites).toHaveBeenCalledTimes(0);
      expect(removeFromFavourites).toHaveBeenCalledWith(mockGame.short);
    });
  });
});
