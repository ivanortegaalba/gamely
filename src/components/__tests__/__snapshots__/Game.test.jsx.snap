// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game/> represent a favourite game should render correctly 1`] = `
<styled.article>
  <header>
    8-Ball Pool
  </header>
  <figure>
    <img
      alt="8-Ball Pool"
      height="60"
      src="http://royal1.midasplayer.com//images/games/eightballpool/eightballpool_60x60.gif"
      width="60"
    />
  </figure>
  <footer>
    <styled.button
      dataSpec="remove-button"
      margin="0 xs 0 0"
      onClick={[Function]}
      palette="danger"
      reverse={true}
      size="s"
    >
      Remove
    </styled.button>
    <styled.button
      dataSpec="play-button"
      margin="0 0 0 xs"
      onClick={[Function]}
      palette="default"
      reverse={true}
      size="s"
    >
      Play
    </styled.button>
  </footer>
</styled.article>
`;

exports[`<Game/> represent a no favourite game should render correctly 1`] = `
<styled.article>
  <header>
    8-Ball Pool
  </header>
  <figure>
    <img
      alt="8-Ball Pool"
      height="60"
      src="http://royal1.midasplayer.com//images/games/eightballpool/eightballpool_60x60.gif"
      width="60"
    />
  </figure>
  <footer>
    <styled.button
      dataSpec="save-button"
      margin="0 xs 0 0"
      onClick={[Function]}
      palette="primary"
      reverse={false}
      size="s"
    >
      Save
    </styled.button>
    <styled.button
      dataSpec="play-button"
      margin="0 0 0 xs"
      onClick={[Function]}
      palette="default"
      reverse={true}
      size="s"
    >
      Play
    </styled.button>
  </footer>
</styled.article>
`;
